### ISSUE TITLE
[BUG] Volume subpath cannot be cleaned up after the node of the workload pod is down and recovered

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#3080]
**Describe the bug**
After the node back, the old pod should be cleaned up without **endless error logs**

**To Reproduce**
1. set up 1 master + 1 worker k3s cluster
2. Launch a 1-replica deployment on the worker node. The containers of the pod use subpath.
3. Kill the k3s agent on the worker node and wait for the pod being terminating.
4. Re-run the k3s agent cmd. The error log will be printed in the worker side:

The deployment YAML file I am using:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mysql-pvc
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: longhorn
  resources:
    requests:
      storage: 5Gi
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysql
  labels:
    app: mysql
  namespace: default
spec:
  selector:
    matchLabels:
      app: mysql
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: mysql
    spec:
      nodeSelector:
        kubernetes.io/hostname: shuo-k3s-01  # worker node
      restartPolicy: Always
      containers:
      - name: mysql
        image: mysql
        command:
        - sh
        - -c
        - "while true; do dd if=/dev/urandom of=/var/lib/mysql/t bs=1M count=1500 oflag=dsync && rm /var/lib/mysql/t; done"
        env:
        - name: MYSQL_ROOT_PASSWORD
          value: "rootpasswd"
        volumeMounts:
        - mountPath: /var/lib/mysql
          name: site-data
          subPath: mysql
      - name: php
        image: php:7.0-apache
        command:
        - sh
        - -c
        - "sleep infinity"
        volumeMounts:
        - mountPath: /var/www/html
          name: site-data
      volumes:
      - name: site-data
        persistentVolumeClaim:
          claimName: mysql-pvc
```

**Expected behavior**
No endless log like `orphaned pod \"a5224f39-1db5-485c-84a9-b4826883e1f5\" found, but failed to rmdir() subpath at path /var/lib/kubelet/pods/a5224f39-1db5-485c-84a9-b4826883e1f5/volume-subpaths/pvc-78c08abc-6814-4110-9ee3-219cf4ba123b/mysql/0: device or resource busy` in k3s/kubelet.

**Log**
```
E0928 12:45:41.088557   22535 kubelet_volumes.go:225] "There were many similar errors. Turn up verbosity to see them." err="orphaned pod \"a5224f39-1db5-485c-84a9-b4826883e1f5\" found, but failed to rmdir() subpath at path /var/lib/kubelet/pods/a5224f39-1db5-485c-84a9-b4826883e1f5/volume-subpaths/pvc-78c08abc-6814-4110-9ee3-219cf4ba123b/mysql/0: device or resource busy" numErrs=1
I0928 12:45:43.088096   22535 kubelet_getters.go:300] "Path does not exist" path="/var/lib/kubelet/pods/a5224f39-1db5-485c-84a9-b4826883e1f5/volumes"
I0928 12:45:43.088127   22535 kubelet_getters.go:300] "Path does not exist" path="/var/lib/kubelet/pods/a5224f39-1db5-485c-84a9-b4826883e1f5/volumes"
I0928 12:45:43.088230   22535 kubelet_volumes.go:140] "Cleaned up orphaned pod volumes dir" podUID=a5224f39-1db5-485c-84a9-b4826883e1f5 path="/var/lib/kubelet/pods/a5224f39-1db5-485c-84a9-b4826883e1f5/volumes"
```

**Environment:**
 - Longhorn version: v1.2.1-rc2
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: v1.21.4+k3s1
   - Number of management node in the cluster: 1
   - Number of worker node in the cluster: 1

 - Number of Longhorn volumes in the cluster: 1

**Comment by shuo-wu** [comment]
Sometimes the error logs mentioned above won't be triggered. But the mount path in the pod is unavailable after the node back and the pod restart. e.g.,
```
$ k exec -it mysql-8495d4d479-gdbnv -- bash
Defaulted container "mysql" out of: mysql, php
root@mysql-8495d4d479-gdbnv:/# ls /var/lib/mysql/
ls: reading directory '/var/lib/mysql/': Input/output error
```

**Comment by shuo-wu** [comment]
Workaround: Scale down then re-scale up the workload.

**Comment by derekbit** [comment]
The NodeUnpublishVolume is not called after kubelet restart sometimes, so the volume is not unmounted.

The race condition is resolved in this commit since kubernetes 1.22.0.

https://github.com/kubernetes/kubernetes/commit/3eadd1a9ead7a009a9abfbd603a5efd0560473cc

**Comment by derekbit** [comment]
The commit is not back ported from kubernetes 1.22 to 1.21 or before.
I'll document it on longhorn.io knowledge base.

**Comment by innobead** [comment]
@khushboo-rancher I know we cover test cases for node down, but do we have this included?

**Comment by khushboo-rancher** [comment]
@innobead You're right, we need to explicitly call this case out of kubelet/k3s agent restart. 
We have automation around the case where we crash instance manager and remount the pod with subpath https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/test_ha.py#L2047 and this is running fine on k8s and k3s env.

We'll add the test case in manual test case during validation of the issue.

**Comment by kaxing** [comment]
Testing in progress.

Trying to reproduce the issue with k3s "v1.21.7+k3s1" + longhorn master, but not really able to see the race condition.
Rebooting the targeting-node a couple more times still seems holding. Will log more notes here if anything.

**Comment by kaxing** [comment]
TL;DR: This is a Kubernetes problem on cleaning up orphaned pod data under volume path. User can prevent this happen with Kuberenetes v1.22 but not able to resolved existed error automatically, even after upgrading the cluster.

---

I can now consistently reproduce this on k3s v1.20.14+k3s1 + Longhorn v1.2.2.
```
Dec 28 15:03:20 c1-worker1 k3s[56124]: I1228 15:03:20.783593   56124 kubelet_getters.go:300] "Path does not exist" path="/var/lib/kubelet/pods/8789bb85-2b5c-4d5f-a0bc-709b467b6626/volumes"
Dec 28 15:03:20 c1-worker1 k3s[56124]: I1228 15:03:20.784091   56124 kubelet_getters.go:300] "Path does not exist" path="/var/lib/kubelet/pods/8789bb85-2b5c-4d5f-a0bc-709b467b6626/volumes"
Dec 28 15:03:20 c1-worker1 k3s[56124]: I1228 15:03:20.784339   56124 kubelet_volumes.go:160] "Cleaned up orphaned pod volumes dir" podUID=8789bb85-2b5c-4d5f-a0bc-709b467b6626 path="/var/lib/kubelet/pods/8789bb85-2b5c-4d5f-a0bc-709b467b6626/volumes"
Dec 28 15:03:20 c1-worker1 k3s[56124]: E1228 15:03:20.784460   56124 kubelet_volumes.go:245] "There were many similar errors. Turn up verbosity to see them." err="orphaned pod \"8789bb85-2b5c-4d5f-a0bc-709b467b6626\" found, but failed to rmdir() subpath at path /var/lib/kubelet/pods/8789bb85-2b5c-4d5f-a0bc-709b467b6626/volume-subpaths/pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4/mysql/0: device or resource busy" numErrs=1
Dec 28 15:03:21 c1-worker1 k3s[56124]: I1228 15:03:21.047035   56124 reconciler.go:224] "operationExecutor.VerifyControllerAttachedVolume started for volume \"pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4\" (UniqueName: \"kubernetes.io/csi/driver.longhorn.io^pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4\") pod \"mysql-5cb7b8fdbd-v6fdv\" (UID: \"bad6e339-5061-4ce0-a62c-7218ecab090f\") "
Dec 28 15:03:21 c1-worker1 k3s[56124]: E1228 15:03:21.047709   56124 nestedpendingoperations.go:335] Operation for "{volumeName:kubernetes.io/csi/driver.longhorn.io^pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4 podName: nodeName:}" failed. No retries permitted until 2021-12-28 15:03:29.04767865 +0800 CST m=+561.393408566 (durationBeforeRetry 8s). Error: Volume has not been added to the list of VolumesInUse in the node's volume status for volume "pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4" (UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4") pod "mysql-5cb7b8fdbd-v6fdv" (UID: "bad6e339-5061-4ce0-a62c-7218ecab090f")
```
However, the volume and pods are appear to be healthy, even after adjust the deployment to 3 replicas:
```
NAME                     READY   STATUS    RESTARTS   AGE
mysql-5cb7b8fdbd-cnnmz   2/2     Running   0          61s
mysql-5cb7b8fdbd-gwpsc   2/2     Running   0          28m
mysql-5cb7b8fdbd-zrpd6   2/2     Running   0          61s
NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE         NODE         AGE
pvc-3b69e9f9-868f-4fd4-a93a-8826569f16c4   attached   healthy      True        5368709120   c1-worker1   37m
```

Adjusted steps to reproduce the problem:
1. Create a cluster with kubernetes v1.21 or before
2. Deploy the example workload, make sure all pods are on the same worker node
    * label the first worker node
    * nodeSelector with the label
3. Stop then start the kubelet agent again with `sudo service k3s-agent restart`
4. Watch the k3s-agent log, orphaned pod error message will present

And the workaround dose not needed in Kubernetes v1.22 to avoid the problem.

**Comment by kaxing** [comment]
Moving this back to analysis, since the upstream fix doesn't really fix in some condition (with 5 replicas I can still reproduce this issue fairly easy)

**Comment by derekbit** [comment]
The subpath deletion failure is related to the cleanup mechanism in `kubelet`.

### **kubernetes 1.21.6**

1. Pod `ADD` events handling
    - `kubelet` on the restarted worker node received a series of pod `ADD` events from the server node.
    - The events are handled by `HandlePodAdditions`. If a pod is at `terminating` state and its `DeletionTimestamp` is set, the containers of the pod is stopped.
    - In this step, the volumes of the terminated pod are not unmounted.
    
2. SyncLoop (housekeeping)
    - Add the pod the pending termination map

3.  Pods killing goroutine received the the notification from step 2
    - `pk.killPod` fails to delete the pod and remove the pod from the  pending termination map.
    
4. SyncLoop DELETE 
    - Pod is fully terminated and removed from etcd by `kl.statusManager.TerminatePod(pod)`

5. SyncLoop REMOVE 
    - `kl.deletePod` tries to delete the pod again
    - The pod is deleted from the pod manager by `kl.probeManager.RemovePod`

6. SyncLoop (housekeeping) 
    - `HandlePodCleanups` cleans up orphaned pod volumes dir
    - **The mounted subpath cannot be deleted by `rmdir()`**
    
### **kubernetes 1.23.3**
The handling logic in the kubelet should avoid the subpath deletion failure.

`kubelet` on the restarted worker node received a series of pod `ADD` events from the server node.
The event is handled by the `HandlePodAdditions`. The terminating pod is terminated by `syncTerminatingPodFn` that stops the running containers in the pod and `syncTerminatedPodFn` that waits for complete unmount of the all volumes by `volumeManager.reconcile`. 
Before unmounting completely, the `SyncLoop (housekeeping)` does not clean the volumes up, because the terminating pod is still in the pod manager.


### Workaround
User can unumount the subpath by manual. Then, the subpath can be deleted by the kubelet successfully.
